sstable: reader: preempt after every fragment

avikivity opened the pull request:
Whenever we push a fragment, we check whether the buffer is
full and return proceed::no if so, so that the state machine pauses
and lets the consumer continue. This patch adds an additional
condition - if preemption is needed, we also return proceed::no.
This drops us back to the outer loop
(in sstable_mutation_reader::fill_buffer), which will yield to
the reactor as part of seastar::do_until().

Two cases (partition_start and partition_end) did not have the
check for is_buffer_full(); it is added now. This can trigger
is the partition has no rows.

Unlike the previous attempt, push_ready_fragments() is not touched.

The extra preemption opportunities triggered a preexisting bug in
clustering_ranges_walker; it is fixed in the first patch of the series.

I tested this by reading from a large partition with a simple
schema (pk int, ck int, primary key(pk, ck)) with BYPASS CACHE.
However, even without the patch I only got sporadic stalls
with the detector set to 1ms, so it's possible I'm not testing
correctly.

Test: unit (dev, debug, release)

Fixes #7883.

avikivity commented:
Changes from v1:
 - added a new patch (patch 1) to address sstable_mutation_test failure in debug mode
 - retested in debug and release modes too (still in progress, but looking good)

tgrabiec commented on sstables/mp_row_consumer.hh:
```diff
@@ -82,6 +82,10 @@ public:
 
+    // Called when a clustering range is exhausted, but more clustering ranges
+    // remain (so a promoted index seek is needed).
+    virtual void on_clustering_range_change() = 0;
```
It seems that you never call it from the ka/la parser.

tgrabiec replied:
An omission like this would manifest in lower performance, since it's always ok to not call advance_context(), we have filtering in place.

May be worthwhile to compare perf_fast_forward diff for blatant regressions.

avikivity replied:
I meant to, somehow I missed it. I'll check with p_f_f.

tgrabiec commented on sstables/mp_row_consumer.hh:
```diff
@@ -974,6 +978,7 @@ class mp_row_consumer_m : public consumer_m {
                 return proceed::no;
             }
             if (_mf_filter->is_current_range_changed()) {
+                _reader->on_clustering_range_change();
```
I think we're missing `on_clustering_range_change()` calls in some cases. E.g. when slicing a partition without a static row we want to immediately skip without reading the front of the partition. Currently, advance_context() will be called before reading from the data file, which will skip.

tgrabiec commented on sstables/partition.cc:
```diff
@@ -433,7 +437,15 @@ class sstable_mutation_reader : public mp_row_consumer_reader {
                     if (is_buffer_full() || _partition_finished || _end_of_stream) {
                         return make_ready_future<>();
                     }
-                    return advance_context(_consumer.maybe_skip()).then([this] {
```
I don't understand why you need this patch. The old code called `maybe_skip()`, which should tell advance_context() to do nothing if skipping wasn't really needed. Maybe the bug was in maybe_skip?

avikivity replied:
Possible, I'll check. My understanding of the code is quite weak.

avikivity replied:
advance_context calls index_reader::advance_to unconditionally:

```c++
    future<> advance_context(std::optional<position_in_partition_view> pos) {
        if (!pos || pos->is_before_all_fragments(*_schema)) {
            return make_ready_future<>();
        }
        assert (_current_partition_key);
        return [this] {
            if (!_index_in_current_partition) {
                _index_in_current_partition = true;
                return get_index_reader().advance_to(*_current_partition_key);
            }
            return make_ready_future();
        }().then([this, pos] {
            return get_index_reader().advance_to(*pos).then([this] {
                index_reader& idx = *_index_reader;
                auto index_position = idx.data_file_positions();
                if (!_context->need_skip(index_position.start)) {
                    return make_ready_future<>();
                }
                return _context->skip_to(idx.element_kind(), index_position.start).then([this, &idx] {
                    _sst->get_stats().on_partition_seek();
                    set_range_tombstone_start_from_end_open_marker(_consumer, *_schema, idx);
                });
            });
        });
    }
```

The first call to advance_to() is conditional, but the second is not.

avikivity replied:
Actually, there's an early return up front. So maybe_skip() should have been enough.

avikivity replied:
> I don't understand why you need this patch. The old code called `maybe_skip()`, which should tell advance_context() to do nothing if skipping wasn't really needed. Maybe the bug was in maybe_skip?

You were right, the regressions in perf_fast_forward disappear with patch 1 removed. But I get a failure in sstable_mutation_tests's test_no_index_reads_when_rows_fall_into_range_boundaries. Maybe it's a false positive?

avikivity replied:
(the regression only appears in debug mode, presumably due to higher chances of preemption)

avikivity replied:
Also appears in release mode with task quota set to 5us. But, it looks to me like a true positive, which I can't explain if maybe_skip() is correct.

tgrabiec replied:
If you share the test log with the failure maybe I can guess something.

avikivity replied:
```c++
            auto before = index_accesses();

            {
                assert_that(ms.make_reader(s, tests::make_permit()))
                    .produces(m1)
                    .produces(m2)
                    .produces_end_of_stream();

                BOOST_REQUIRE_EQUAL(index_accesses(), before);
            }
```

index_accesses() = 2, before == 0

Later I'll generate a trace-level log, see if I find anything, and post it here.

avikivity commented:
The tests do pass.

avikivity commented:
perf_fast_forward does show an increase in aio in some cases:

(even lines are before, odd lines are after):

```
running: large-partition-forwarding on dataset large-part-ds1
Testing forwarding with clustering restriction in a large partition:
pk-scan   time (s)   iterations     frags     frag/s    mad f/s    max f/s    min f/s    avg aio    aio      (KiB) blocked dropped  idx hit idx miss  idx blk    c hit   c miss    c blk    cpu 
yes       0.000581         1003         2       3441         35       3950        486        4.0      4        133       3       1        0        1        1        0        0        0  14.5% 
yes       0.000693          688         2       2887          9       2930       1368        6.0      6        197       4       2        0        1        1        0        0        0  13.7% 
no        0.000574         1713         2       3482         74       3922        497        4.0      4        133       3       1        0        1        1        0        0        0  14.4% 
no        0.000692         1420         2       2888         15       2960        483        6.0      6        197       4       2        0        1        1        0        0        0  12.9% 
```

and

```
running: large-partition-slicing-clustering-keys on dataset large-part-ds1
Testing slicing of large partition using clustering keys:
offset  read      time (s)   iterations     frags     frag/s    mad f/s    max f/s    min f/s    avg aio    aio      (KiB) blocked dropped  idx hit idx miss  idx blk    c hit   c miss    c blk    cpu 
500000  1         0.000520         1696         1       1924          3       2071        255        4.0      4        133       3       1        0        1        1        0        0        0  13.0% 
500000  1         0.000690         1142         1       1450          4       1509        239        6.0      6        197       4       2        0        1        1        0        0        0  11.4% 
500000  32        0.000523         1854        32      61197        194      61851       8206        4.0      4        133       3       1        0        1        1        0        0        0  14.9% 
500000  32        0.000692         1189        32      46256        101      46584       7666        6.0      6        197       4       2        0        1        1        0        0        0  12.7% 
500000  256       0.000709         1380       256     361134       1250     364805      61262        5.0      5        165       4       1        0        1        1        0        0        0  20.6% 
500000  256       0.000885          827       256     289160        717     290734     221382        7.0      7        229       5       2        0        1        1        0        0        0  18.3% 
500000  4096      0.002823          424      4096    1451059      85137    2222133     748942       15.2     18        581      17       1        0        1        1        0        0        0  44.7% 
500000  4096      0.003066          329      4096    1335927      43031    2125532     659387       19.6     20        645      18       2        0        1        1        0        0        0  41.0% 

```

avikivity commented:
```TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - index 0x601000141a40: index_reader for /tmp/scylla-fc9fdc0b-6a2a-4fd4-87a4-bf67d33d3b20/mc-1-big-Data.db
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - index 0x601000141a40: advance_to_end() bound 0x601000141b08
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - reader 0x6010001edf00: reading partition
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - reader 0x6010001edf00: read from data file
TRACE 2021-01-26 12:26:27,040 [shard 0] io - dev 0 : req 0x6010001f1500 queue  len 4096 ticket 128:1024
TRACE 2021-01-26 12:26:27,040 [shard 0] io - dev 0 : req 0x6010001f1500 submit
TRACE 2021-01-26 12:26:27,040 [shard 0] io - dev 0 : req 0x6010001f1500 complete
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_partition_start(deltime=(2147483647, -9223372036854775808)), _is_mutation_end=true
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: setup_for_partition(pk{000401000000})
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_static_row_start()
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_end(_in_progress_static_row={static_row: {row: }})
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303031})
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303031},0}
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775807, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - index 0x601000141a40: advance_to({position: clustered,ckp{},-1}), current data_file_pos=0
TRACE 2021-01-26 12:26:27,040 [shard 0] sstable - index 0x601000141a40: advance_to_page(0), bound 0x601000141aa0
TRACE 2021-01-26 12:26:27,041 [shard 0] io - dev 0 : req 0x6010001f1500 queue  len 4096 ticket 128:1024
TRACE 2021-01-26 12:26:27,041 [shard 0] io - dev 0 : req 0x6010001f1500 submit
TRACE 2021-01-26 12:26:27,041 [shard 0] io - dev 0 : req 0x6010001f1500 complete
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 0 state START
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: exit pos 0 state KEY_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 0 state KEY_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 2 state KEY_BYTES
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 6 state POSITION
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 7 state PROMOTED_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 8 state PARTITION_HEADER_LENGTH_1
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 8 state CONSUME_ENTRY size 0
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: exit pos 8 state START
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 8 state START
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: exit pos 8 state KEY_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 8 state KEY_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 10 state KEY_BYTES
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 14 state POSITION
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 15 state PROMOTED_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 16 state PARTITION_HEADER_LENGTH_1
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 16 state CONSUME_ENTRY size 0
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: exit pos 16 state START
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: pos 16 state START
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - 0x601000306e38: exit pos 16 state KEY_SIZE
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - index 0x601000141a40 bound 0x601000141aa0: page:
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable -   {key: pk{000401000000}, token:-8604791237420463362} -> 0
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable -   {key: pk{000400000000}, token:-3485513579396041028} -> 108
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - index 0x601000141a40: page done
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - index 0x601000141a40: advance_to({position: clustered,ckp{},-1}), current data_file_pos=0
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - index 0x601000141a40: no promoted index
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303032})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303032},0}
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775806, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303035})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303035},0}
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775805, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303036})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303036},0}
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775804, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_partition_end()
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_partition_start(deltime=(2147483647, -9223372036854775808)), _is_mutation_end=true
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: setup_for_partition(pk{000400000000})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_static_row_start()
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=7376616c7565, ts=-9223372036854775803, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_end(_in_progress_static_row={static_row: {row: 
      {s1, { atomic_cell{svalue,ts=-9223372036854775803,expiry=-1,ttl=0} }}}})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303032})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303032},0}
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775802, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303035})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303035},0}
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775801, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303036})
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303036},0}
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775800, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_partition_end()
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - reader 0x6010001edf00: reading partition
TRACE 2021-01-26 12:26:27,041 [shard 0] sstable - reader 0x6010001edf00: read from data file
TRACE 2021-01-26 12:26:27,041 [shard 0] testlog - Expecting end of stream
test/boost/sstable_mutation_test.cc(1372): fatal error: in "test_no_index_reads_when_rows_fall_into_range_boundaries": critical check index_accesses() == before has failed [2 != 0]
```

avikivity commented:
I added trace messages to mutation_fragment_filter::maybe_skip:

```
RACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303032})
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303032},0}
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775806, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - maybe_skip: current range unchanged, no skip
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303035})
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303035},0}
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775805, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - maybe_skip: current range unchanged, no skip
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303036})
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: emit
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303036},0}
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775804, ttl=0, del_time=9223372036854775807, deleted=false)
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - maybe_skip: current range unchanged, no skip
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_partition_end()
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - reader 0x6010001edf00: reading partition
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - reader 0x6010001edf00: read from data file
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: consume_partition_start(deltime=(2147483647, -9223372036854775808)), _is_mutation_end=true
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - mp_row_consumer_m 0x6010001edfb0: setup_for_partition(pk{000400000000})
TRACE 2021-01-26 12:37:28,493 [shard 0] sstable - maybe_skip: {position: static_row,null,0}
```

avikivity commented:
It is odd that maybe_skip() returned a positive immediately after setup_for_partition(), which initializes _mf_filter and the associated clustering_range_walker.

avikivity commented:
I added more logging. Diff between a bad and a good run:

```
+sstable - is_current_range_changed: 0 1
+sstable - maybe_skip: {position: static_row,null,0}
```

We got preempted, and maybe_skip told us to skip in the bad run, but we didn't go to the index (likely since this is the beginning of the partition).

```
 sstable - mp_row_consumer_m 0x6010001edfb0: consume_static_row_start()
 sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_end(_in_progress_static_row={static_row: {row: }})
+sstable - is_current_range_changed: 1 1
+sstable - maybe_skip: current range unchanged, no skip
```

Preempted after the static row, correctly told not to skip.

```
 sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_start(ckp{000c636b30303030303030303031})
 sstable - clustering_range_walker: advance_to (entry)
 sstable - clustering_range_walker: advance_to_next_range (entry)
@@ -598,215 +602,75 @@
 sstable - mp_row_consumer_m 0x6010001edfb0: emit
 sstable - mp_row_consumer_m 0x6010001edfb0: consume_row_marker_and_tombstone({row_marker: }, {tombstone: none}, {tombstone: none}), key={position: clustered,ckp{000c636b30303030303030303031},0}
 sstable - mp_row_consumer_m 0x6010001edfb0: consume_column(id={0}, path=, value=76, ts=-9223372036854775807, ttl=0, del_time=9223372036854775807, deleted=false)
+sstable - is_current_range_changed: 1 2
+sstable - maybe_skip: {position: clustered,ckp{},-1}
```

Preempted after the first clustering row, told to skip.

```
+sstable - index 0x601000141a40: advance_to({position: clustered,ckp{},-1}), current data_file_pos=0
+sstable - index 0x601000141a40: advance_to_page(0), bound 0x601000141aa0
```

That's the skip.

avikivity commented:
This is the culprit:

```c++
            , _current_start(position_in_partition_view::for_static_row())
            , _current_end(position_in_partition_view::before_all_clustered_rows()) {
```

current_end needlessly splits the infinite clustering range into two.

avikivity commented:
Victory! (unless I caused 82,112 new regressions)

avikivity commented:
Updated:
 - dropped patch to communicate skips from parser to sstable_mutation_reader, since mutation_fragment_filter::maybe_skip() already does that
 -  fixed bug in clustering_ranges_walker, where a static row and a clustering interval with -inf as its left bound were processed as two separate ranges (new patch 1)

avikivity commented:
The new tests failed in debug mode, since clustering_ranges_walker accepts its ranges argument by reference, and I supplied it by value. I dequeued it. The fix is trivial, but will retest everything to avoid more embarrassment.

avikivity commented:
Updated:
 - rebased
 - changed clustering_ranges_walker ranges parameter to not be a temporary
 - retested (debug, dev, release)
